### PR TITLE
Enable Rubin MXFP8 clamped GeGLU forward and backward

### PR DESCRIPTION
**Pull request by ZeYang1025**
## Before submitting

- [x] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt).
- [x] I ran `pre-commit run` and committed any formatting changes.
- [x] I reviewed the Hard Rules in the `AGENTS.md` for each directory this PR touches (see [root AGENTS.md § Reviewing a PR](https://github.com/NVIDIA/cudnn-frontend/blob/develop/AGENTS.md#reviewing-a-pr-human-or-agent)) and my changes comply, or I explain the exception below.
- [ ] I added GitHub labels: one `cat-*`, one or more `area:*` / `op:*`, and one `orig-*` (see [label list](https://github.com/NVIDIA/cudnn-frontend/labels)).

Suggested labels: `cat-feature`, `cat-bugfix`, `mod-cutedsl`, `orig-nv-eng`. Label assignment requires maintainer access; the repository currently has no grouped-GEMM `area:*` / `op:*` label.

## Affected area

FE OSS kernels or CuTeDSL

## Summary

Enable configurable GeGLU alpha and clamp bounds in Rubin grouped GEMM forward/backward, allowing DeepSeek V4 Flash clamped SwiGLU with MXFP8 to execute through the fused expert path. Cover dense/discrete weights and scalar/packed epilogues, fix missing GEMM `alpha_tensor²` scaling in backward, and update the existing API documentation and tests.

## Why

DeepSeek V4 requires activation alpha `1`, linear offset `0`, and clamp bounds `[-10, 10]`. Rubin previously rejected these settings and used fixed alpha/clamp constants. Its backward GeGLU path also omitted the operand-scaling factor needed for activation and routing-probability gradients when GEMM alpha is nonunit.

## Related issues

No linked issue.

## API and compatibility impact

Existing public signatures and defaults are preserved. Forward activation parameters remain runtime values and reuse the compiled kernel; backward parameters retain the existing configuration specialization and cache behavior. The gate is upper-clamped only, the up branch is clamped on both sides, and stored C remains unclipped. Clamp-boundary derivatives follow the existing Blackwell convention.

## Testing

On Rubin SM107, the following pytest selection ran through a harness that verified the loaded checkout sources:

```bash
cd test/python
pytest -s -p no:xdist -v --tb=long -m 'L0 or L1' -k rubin_mxfp8 \
  fe_api/grouped_gemm/test_grouped_gemm_glu.py \
  fe_api/grouped_gemm/test_grouped_gemm_dglu.py
```

- **17 passed; zero failures, errors, or skips.** Independent references check clamp boundaries, defaults/custom parameters, probabilities, nonunit GEMM alpha, output quantization/scales, AMAX, and cache behavior. The unpatched baseline reproduced both unsupported-parameter guards and the backward nonunit-alpha numerical failure. Two neighboring BF16 validation tests also passed in a separate run.
- Real Transformer Engine experts passed forward/backward with finite outputs and gradients. Megatron-Bridge's DeepSeek V4 Flash MXFP8 library recipe completed **two unskipped optimizer steps on four GPUs**, using a **four-layer prefix, original expert widths/count, sequence length 512, mock data, and random initialization**. All four expert layers on every rank selected fused forward/backward with the DSv4 settings; losses and gradient norms were finite.
- Validation used FE 1.30 Python operator sources over existing FE 1.25 native bindings, CUDA 13.5, and CuTe DSL 4.8.0.dev0. This was a verified source overlay, not a rebuilt FE package. Full-depth/64-GPU training, Blackwell execution, long-run convergence, and performance were not tested.
- `pre-commit run`: Black 26.3.1, Black Jupyter, and SPDX checks passed; no applicable clang-format files. Python syntax and `git diff --check` also passed.
- The dGLU reference explicitly preserves FE's inclusive clamp-boundary gradients across PyTorch versions. [PyTorch 2.14 changed scalar-bound `clamp` gradients at equality from 1 to 0](https://github.com/pytorch/pytorch/pull/191142). On SM107, a diagnostic applying that strict boundary rule reproduced the two CI failures exactly (386 and 21 scale mismatches); the corrected reference passed all 17 tests with native PyTorch 2.13 behavior and all 17 with the strict-rule diagnostic. The private CI image requires registry access, so the exact CI stack still needs a rerun. Kernel code, tolerances, and test count are unchanged by this correction.
- The training validation used baseline `94939f3a` plus this patch. Operator files remain byte-identical to that tested version; the reference-only correction was validated separately as described above.

## Summary by CodeRabbit

- **New Features**
  - Added Rubin SM107 support for block-scaled GLU and dGLU operations.
  - Added block-scaled Rubin dSwiGLU and dGeGLU workflows.
  - Added runtime configuration for GeGLU activation alpha, clamp limits, and linear offset across dense and discrete paths.
  - Added compilation and execution support for customized Rubin activation settings.

- **Documentation**
  - Documented supported architectures, activation controls, defaults, scaling behavior, gradient handling, environment requirements, and MXFP8 constraints.

### CODE REVIEW

**Review by YangXu1990uiuc — APPROVED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `7dd3597aeadc` against base `b03c3f3321f5`.

Thank you for your contribution! The Rubin forward/backward parameter plumbing, GEMM-alpha scaling, clamp derivatives, and cache behavior look sound at 7dd3597aeadc. I found no approval-blocking issue. The new tests cover both scalar and packed math, dense and discrete weights, runtime changes, quantized outputs, and zero routing probabilities with nonzero dprob. One small documentation correction below.

**[P3] Point the Rubin example reference to the existing test module** ([source](https://github.com/NVIDIA/cudnn-frontend/blob/7dd3597aeadc2cfca35770eb3261b5e25cc033d8/docs/fe-oss-apis/gemm_fusions/grouped_gemm_dglu.md#L653))

The documented test_grouped_gemm_dglu_rubin_mxfp8.py does not exist in this revision. The new Rubin regressions are in test_grouped_gemm_dglu.py (test_rubin_mxfp8_clamped_dgeglu_*). Please update this path so readers can find and run the advertised examples.

Evidence: Checked the repository paths and the new test definitions at the reviewed head. Attribution: Introduced by this PR's documentation addition.

Validation: Independent exact-head Rubin SM107 run: 17 passed, 1488 deselected, zero skips across the new GLU/dGLU regressions, explicitly selecting L0 or L1. Verified that both APIs selected their Rubin kernels; the tests include independent activation/autograd references, quantized payload/scales, AMAX, dprob, and cache reuse/reconfiguration. Independent SM100 neighboring regression run: 14 passed, 1491 deselected, zero skips for existing discrete alpha/clamp cases and BF16 scale-override rejection. Built the native Python binding from this exact head; validation used CUDA 13.2, cuDNN 9.25.1, and CuTe DSL 4.8 development dependencies. Reviewed the public API signatures and both dense/discrete compile and execute paths.

Limitations: A before/after Rubin CPU-enqueue/GPU timing probe was prepared but did not start: after the completed correctness suite, the device health check reported an NVML device error. No performance result or attribution to this PR is claimed. The broader current-head CI has not yet run. Style/Docs require fork workflow approval; the focused hardware results above do not substitute for the full suite or a training integration run.

Approved: no P0 or high-risk P1 found. Remaining findings stay with the owner; merge timing stays with the owner.

**Comment by YangXu1990uiuc** [comment]
I am requesting test CI for `7dd3597aeadc` with the following standalone command: `@cudnn-ci-bot run oss`. For future updates, request the appropriate test target before review; a maintainer may need to trigger it.

Codex bot review — model `gpt-6-astra`.

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run oss

**Comment by ZeYang1025** [comment]
Thanks for running CI and sharing the logs! The failures came from the test reference: PyTorch 2.14 changed `torch.clamp` gradients at exact clamp boundaries from 1 to 0, while FE uses an inclusive boundary convention.

I’ve updated the reference to preserve that convention explicitly with `torch.where` and pushed the fix in `6257fb35`.

I’m still getting familiar with the CI workflow. Please let me know if there’s anything I should do to trigger a rerun. Thanks for your help!

**Review by YangXu1990uiuc — COMMENTED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `6257fb35da1d` against base `b03c3f3321f5`.

Thank you for your contribution! Rechecked the incremental change through 6257fb35da1d: the documentation path is fixed, and the reference now explicitly preserves the documented inclusive clamp-boundary derivative. No remaining review findings; the existing approval remains applicable. The production API and kernels are unchanged from the previously reviewed revision. Current-head Rubin CI is still needed to confirm the reported CI-environment reference failure is resolved.

Validation: Ran the actual current-head rubin_mxfp8_dglu_reference against an independent closed-form derivative: 6 cases passed (dense/discrete x three parameter sets), covering exact lower/upper boundaries, out-of-range inputs, nonunit GEMM alpha, zero routing probability, dC, dprob and the reference error scale. A deliberately exclusive-boundary oracle differs substantially, so the cases distinguish the intended convention. Confirmed only documentation and the test reference changed since 7dd3597aeadc. The prior independent 17-case Rubin run and 14-case SM100 neighboring-path run remain evidence for the unchanged production code; they are not a current-head CI result. Checked the corrected documentation target and test-name prefix against the current source.

Limitations: The new reference probe used PyTorch 2.13 on SM100 and executed no FE GPU kernel. It verifies the explicit derivative contract, not the PyTorch 2.14 CI environment; the author's attribution of the earlier Rubin CI failure awaits confirmation by the rerun. No new performance comparison was needed for this reference/documentation-only delta. The prior review did not establish a Rubin performance result.

Comment only: this PR already has an active approval; preserving existing approvals.

**Comment by YangXu1990uiuc** [comment]
The incremental reference fix has been reviewed, including an independent analytic check of the inclusive clamp-boundary derivatives. I am requesting `oss` CI for `6257fb35da1d` and approving its pending Style/Docs test workflows.

Your trigger attempt was blocked by the CI allowlist. For future updates, ask an allowlisted maintainer to post the standalone command `@cudnn-ci-bot run oss`; no additional change to your PR is needed to obtain that permission for this run.

Codex bot review — model `gpt-6-astra`.

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run oss

**Comment by YangXu1990uiuc** [comment]
CI follow-up for `6257fb35da1d`: the requested `oss` pipeline **68701573 passed all 24 executed jobs**, including the previously failing Rubin OSS lane and both SM100 shards. Style and Docs also passed. The six other manual target groups were not requested.

This closes the pending CI check from my follow-up review; the existing approval remains applicable.

Codex bot review — model `gpt-6-astra`.

**Review by Anerudhan — APPROVED** [review]
